coordinator-common/src/main/java/org/apache/kafka/coordinator/common/runtime/CoordinatorMetadataImage.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.common.runtime;
+
+import org.apache.kafka.common.Uuid;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Provides metadata to Coordinators (GroupCoordinator, ShareCoordinator, etc) such as topics, partitions, and their configurations.
+ * Implementations should be thread-safe and immutable.
+ */
+public interface CoordinatorMetadataImage {
+    CoordinatorMetadataImage EMPTY = emptyImage();
+
+    Optional<String> topicName(Uuid id);
+
+    Optional<Uuid> topicId(String topicName);
+
+    default Optional<Integer> partitionCount(Uuid topicId) {
+        var topicName = topicName(topicId);
+        return topicName.isEmpty() ? Optional.empty() : partitionCount(topicName.get());
+    }
+
+    Optional<Integer> partitionCount(String topicName);
+
+    Set<Uuid> topicIds();
+
+    Set<String> topicNames();
+
+    default Optional<TopicMetadata> topicMetadata(String topicName) {
+        var topicId = topicId(topicName);
+        return topicId.isEmpty() ? Optional.empty() : topicMetadata(topicId.get());
+    }
+
+    Optional<TopicMetadata> topicMetadata(Uuid topicId);
+
+    boolean shareGroupsEnabled();
```

Comment:
This feels weird in the interface. I wonder whether we could expose the features instead. It seems useful for other needs too.

Reply:
The thing is that the decision whether share groups are enabled depends upon a config and a feature. Either can be used.